src/librbd/io/ReadResult.cc

```diff
@@ -191,26 +189,49 @@ void ReadResult::C_ObjectReadRequest::finish(int r) {
     r = 0;
   }
   if (r >= 0) {
-    ldout(cct, 10) << " got " << extent_map
-                   << " for " << buffer_extents
-                   << " bl " << bl.length() << dendl;
+    uint64_t object_len = 0;
     aio_completion->lock.lock();
-    if (!extent_map.empty()) {
+    for (auto& extent: extents) {
+      ldout(cct, 10) << " got " << extent.extent_map
+                     << " for " << extent.buffer_extents
+                     << " bl " << extent.bl.length() << dendl;
+
       aio_completion->read_result.m_destriper.add_partial_sparse_result(
-        cct, bl, extent_map, object_off, buffer_extents);
-    } else {
-      // handle the case where a sparse-read wasn't issued
-      aio_completion->read_result.m_destriper.add_partial_result(
-        cct, std::move(bl), buffer_extents);
+              cct, extent.bl, extent.extent_map, extent.offset,
```

Comment:
If `extent_map` is empty (because it wasn't a sparse-read), shouldn't it just `emplace_back(extent.offset, extent.length)`? I know you are covering this (internally) within `add_partial_sparse_result`, but looking at the unit test cases where the `extent_map` is empty for non-sparse reads seems to be leaking internal implementation details.

Reply:
I don't understand. You're saying that if `extent_map` is empty then no need to call add_partial_result ?
And instead emplace_back to what?

Reply:
I am just saying that it's awkward that `extent.extent_map` is sometimes empty and sometimes populated (based on the read size or the OSD). It can always invoke `add_partial_sparse_result`.

Reply:
You're right, extent_map can be empty or not. And this needs to be checked.
Before this PR, it was checked in ReadResult. I moved it to Striper to be able to re-use this check from other places.
Where do you suggest to put this check?